<a href="https://colab.research.google.com/github/ClaudiOmatheuuss/introducao_pandas/blob/main/Python_para_Data_Science_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <font color=green> PYTHON PARA DATA SCIENCE - PANDAS
---

# <font color=green> 1. INTRODUÇÃO AO PYTHON
---

# 1.1 Introdução

> Python é uma linguagem de programação de alto nível com suporte a múltiplos paradigmas de programação. É um projeto *open source* e desde seu surgimento, em 1991, vem se tornando uma das linguagens de programação interpretadas mais populares.
>
> Nos últimos anos Python desenvolveu uma comunidade ativa de processamento científico e análise de dados e vem se destacando como uma das linguagens mais relevantes quando o assundo é ciência de dados e machine learning, tanto no ambiente acadêmico como também no mercado.

# 1.2 Instalação e ambiente de desenvolvimento

### Instalação Local

### https://www.python.org/downloads/
### ou
### https://www.anaconda.com/distribution/

### Google Colaboratory

### https://colab.research.google.com

### Verificando versão

# 1.3 Trabalhando com dados

In [ ]:
#pd.set_option('display.max_rows', 10) <- permite ver determinada quantidade de linhas
#pd.set_option('display.max_columns', 10) <- permite ver determinada quantidade de colunas

import pandas as pd   # importa biblioteca pandas

In [ ]:
# lê o arquivo .csv externo, transforma em um Dataframe e armazena na variavel dataset
dataset = pd.read_csv('/db.csv', sep = ';')

In [ ]:
dataset

Nome                  Motor   Ano  Quilometragem  \
0            Jetta Variant        Motor 4.0 Turbo  2003        44410.0   
1                   Passat           Motor Diesel  1991         5712.0   
2                 Crossfox        Motor Diesel V8  1990        37123.0   
3                      DS5        Motor 2.4 Turbo  2019            NaN   
4         Aston Martin DB4        Motor 2.4 Turbo  2006        25757.0   
..                     ...                    ...   ...            ...   
253           Phantom 2013               Motor V8  2014        27505.0   
254  Cadillac Ciel concept               Motor V8  1991        29981.0   
255             Classe GLK  Motor 5.0 V8 Bi-Turbo  2002        52637.0   
256       Aston Martin DB5           Motor Diesel  1996         7685.0   
257                  Macan        Motor Diesel V6  1992        50188.0   

     Zero_km                                         Acessórios      Valor  
0      False  ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
1      False  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
2      False  ['Piloto automático', 'Controle de estabilidad...   72832.16  
3       True  ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
4      False  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10  
..       ...                                                ...        ...  
253    False  ['Controle de estabilidade', 'Piloto automátic...   51759.58  
254    False  ['Bancos de couro', 'Painel digital', 'Sensor ...   51667.06  
255    False  ['Rodas de liga', 'Controle de tração', 'Câmbi...   68934.03  
256    False  ['Ar condicionado', '4 X 4', 'Câmbio automátic...  122110.90  
257    False  ['Central multimídia', 'Teto panorâmico', 'Vid...   90381.47  

[258 rows x 7 columns]

In [ ]:
dataset.dtypes   # retorna os tipos de dados de cada coluna do DataFrame

Nome              object
Motor             object
Ano                int64
Quilometragem    float64
Zero_km             bool
Acessórios        object
Valor            float64
dtype: object

In [ ]:
dataset[['Quilometragem', 'Valor']].describe()   # gera um conjunto de estatísticas descritivas das colunas selecionadas de um DataFrame

Quilometragem          Valor
count     197.000000     258.000000
mean    58278.421320   98960.513101
std     35836.733259   29811.932305
min       107.000000   50742.100000
25%     27505.000000   70743.512500
50%     55083.000000   97724.380000
75%     90495.000000  124633.302500
max    119945.000000  149489.920000

In [ ]:
dataset.info() # retorna as informações sobre um DataFrame

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 258 entries, 0 to 257
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Nome           258 non-null    object 
 1   Motor          258 non-null    object 
 2   Ano            258 non-null    int64  
 3   Quilometragem  197 non-null    float64
 4   Zero_km        258 non-null    bool   
 5   Acessórios     258 non-null    object 
 6   Valor          258 non-null    float64
dtypes: bool(1), float64(2), int64(1), object(3)
memory usage: 12.5+ KB


# <font color=green> 2. TRABALHANDO COM TUPLAS
---

# 2.1 Criando tuplas

Tuplas são sequências imutáveis que são utilizadas para armazenar coleções de itens, geralmente heterogêneos. Podem ser construídas de várias formas:
```
- Utilizando um par de parênteses: ( )
- Utilizando uma vírgula à direita: x,
- Utilizando um par de parênteses com itens separados por vírgulas: ( x, y, z )
- Utilizando: tuple() ou tuple(iterador)
```

In [ ]:
()     # tupla vazia

()

In [ ]:
1, 2, 3        # cria tupla com esses valores, uma vez que, o python interpreta assim essa sequência de valores

(1, 2, 3)

In [ ]:
nome = 'Passat'
valor = 153000
(nome, valor)

('Passat', 153000)

In [ ]:
nomes_carros = tuple(['Jetta Variant', 'Passat', 'Crossfox', 'DS5'])
nomes_carros

('Jetta Variant', 'Passat', 'Crossfox', 'DS5')

In [ ]:
type(nomes_carros)

tuple

# 2.2 Seleções em tuplas

In [ ]:
nomes_carros = ('Jetta Variant', 'Passat', 'Crossfox', 'DS5')
nomes_carros

('Jetta Variant', 'Passat', 'Crossfox', 'DS5')

In [ ]:
nomes_carros[0]

'Jetta Variant'

In [ ]:
nomes_carros[1]

'Passat'

In [ ]:
nomes_carros[-1]

'DS5'

In [ ]:
nomes_carros[:-1]

('Jetta Variant', 'Passat', 'Crossfox')

In [ ]:
nomes_carros[1:3]

('Passat', 'Crossfox')

In [ ]:
nomes_carros[-4:-1]

('Jetta Variant', 'Passat', 'Crossfox')

In [ ]:
nomes_carros = ('Jetta Variant', 'Passat', 'Crossfox', 'DS5', ('Fusca', 'Gol', 'C4'))
nomes_carros

('Jetta Variant', 'Passat', 'Crossfox', 'DS5', ('Fusca', 'Gol', 'C4'))

In [ ]:
nomes_carros[-1]

('Fusca', 'Gol', 'C4')

In [ ]:
nomes_carros[-1][1]    # acessa o elemento de index 1 dentro do elemento de posição 4/-1

'Gol'

# 2.3 Iterando em tuplas

In [ ]:
nomes_carros = ('Jetta Variant', 'Passat', 'Crossfox', 'DS5')
nomes_carros

('Jetta Variant', 'Passat', 'Crossfox', 'DS5')

In [ ]:
for item in nomes_carros:
  print(item)

Jetta Variant
Passat
Crossfox
DS5


### Desempacotamento de tuplas

Com o desempacotamento de tuplas é possível distribuir valores de uma tupla e atribuir cada uma a variáveis em específico.

É bastante útil quando temos funções que retornam tuplas e queremos utilizar esses valores para coisas diferentes no código.

In [ ]:
nomes_carros = ('Jetta Variant', 'Passat', 'Crossfox', 'DS5')
nomes_carros

('Jetta Variant', 'Passat', 'Crossfox', 'DS5')

In [ ]:
carro_1, carro_2, carro_3, carro_4 = nomes_carros

In [ ]:
carro_1

'Jetta Variant'

In [ ]:
carro_2

'Passat'

In [ ]:
carro_3

'Crossfox'

In [ ]:
carro_4

'DS5'

In [ ]:
_, A, _, B = nomes_carros    # usando _ podemos ignorar o valor assim pegando apenas o que for necessário do retorno

In [ ]:
A

'Passat'

In [ ]:
B

'DS5'

In [ ]:
_, C, *_ = nomes_carros   # guarda apenas o segundo valor em C e ignora o restante da tupla

In [ ]:
C

'Passat'

## *zip()*    <-  Cria iterador com tuplas

https://docs.python.org/3.6/library/functions.html#zip

In [ ]:
carros = ['Jetta Variant', 'Passat', 'Crossfox', 'DS5']
carros

['Jetta Variant', 'Passat', 'Crossfox', 'DS5']

In [ ]:
valores = [88078.64, 106161.94, 72832.16, 124549.07]
valores

[88078.64, 106161.94, 72832.16, 124549.07]

In [ ]:
list(zip(carros, valores))

[('Jetta Variant', 88078.64),
 ('Passat', 106161.94),
 ('Crossfox', 72832.16),
 ('DS5', 124549.07)]

In [ ]:
"""
      associa as posições dos elementos iterando as duas listas ao mesmo tempo e criando uma tupla para os elementos associados por sua indexação
                                    zip(carros,valores)  ->  (carros[0],valores[0]), (carros[1],valores[1]), (carros[2],valores[2]), (carros[3],valores[3])
"""
for item in zip(carros, valores):
  print(item)

('Jetta Variant', 88078.64)
('Passat', 106161.94)
('Crossfox', 72832.16)
('DS5', 124549.07)


In [ ]:
for carro, valor in zip(carros, valores): # desempacotamento dos valores nas tuplas dentro do metodo zip
  print(carro, valor)

Jetta Variant 88078.64
Passat 106161.94
Crossfox 72832.16
DS5 124549.07


In [ ]:
for carro, valor in zip(carros, valores):
  if (valor > 100000):   # printa todos carros com valor maior que 100.000
    print(carro)

Passat
DS5


# ATIVIDADES

# 1) Para responder esta questão, considere a tupla:


In [ ]:
carros = (
    (
        'Jetta Variant',
        'Motor 4.0 Turbo',
        2003,
        False,
        ('Rodas de liga', 'Travas elétricas', 'Piloto automático')
    ),
    (
        'Passat',
        'Motor Diesel',
        1991,
        True,
        ('Central multimídia', 'Teto panorâmico', 'Freios ABS')
    )
)

 O que precisamos é iterar na tupla carros e imprimir todos os acessórios que aparecem. O resultado desejado é o seguinte:
- Rodas de liga
- Travas elétricas
- Piloto automático
- Central multimídia
- Teto panorâmico
- Freios ABS

In [ ]:
for carro in carros:
  for acessorio in carro[-1]:
    print(acessorio)

Rodas de liga
Travas elétricas
Piloto automático
Central multimídia
Teto panorâmico
Freios ABS


# 2) Considerando as duas listas abaixo:

In [ ]:
nomes = ['Passat', 'Crossfox', 'DS5', 'C4', 'Jetta']

kms = [15000, 12000, 32000, 8000, 50000]

E utilizando a função zip e desempacotamento de tuplas, faça o código que possibilita a impressão dos nomes dos veículos com quilometragem abaixo de 20.000 km.

In [ ]:
for nome,km in zip(nomes,kms):
  if (km < 20000):
    print(nome)

Passat
Crossfox
C4


# <font color=green> 3. TRABALHANDO COM DICIONÁRIOS
---

# 3.1 Criando dicionários

Listas são coleções sequenciais, isto é, os itens destas sequências estão ordenados e utilizam índices (números inteiros) para acessar os valores.

Os dicionários são coleções um pouco diferentes. São estruturas de dados que representam um tipo de mapeamento. Mapeamentos são coleções de associações entre pares de valores onde o primeiro elemento do par é conhecido como chave (*key*) e o segundo como valor (*value*).

```
dicionario = {key_1: value_1, key_2: value_2, ..., key_n: value_n}
```

https://docs.python.org/3.6/library/stdtypes.html#typesmapping

In [ ]:
carros = ['Jetta Variant', 'Passat', 'Crossfox']
carros

['Jetta Variant', 'Passat', 'Crossfox']

In [ ]:
valores = [88078.64, 106161.94, 72832.16]
valores

[88078.64, 106161.94, 72832.16]

In [ ]:
carros.index('Passat')

1

In [ ]:
valores[carros.index('Passat')]

106161.94

In [ ]:
#         = { 'Chave1': valor1,               Chave2': valor2,         Chave3': valor3    }
dados = {'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}
dados

{'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}

In [ ]:
type(dados)

dict

In [ ]:
# em um dicionario os valores são acessados por suas chaves
dados['Passat']

106161.94

### Criando dicionários com *zip()*

In [ ]:
list(zip(carros, valores))

[('Jetta Variant', 88078.64), ('Passat', 106161.94), ('Crossfox', 72832.16)]

In [ ]:
dados = dict(zip(carros, valores))
dados

{'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}

# 3.2 Operações com dicionários

In [ ]:
dados = {'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}
dados

{'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}

## *dict[ key ]*

Retorna o valor correspondente à chave (*key*) no dicionário.

In [ ]:
dados['Passat']

106161.94

## *key in dict*

Retorna **True** se a chave (*key*) for encontrada no dicionário.

In [ ]:
'Passat' in dados

True

In [ ]:
'Fusca Rebaixado' in dados

False

In [ ]:
'Fusca Rebaixado' not in dados

True

## *len(dict)*

Retorna o número de itens do dicionário. Ou seja, quantidade de pares chave-valor presentes no dicionário.

In [ ]:
len(dados)

3

## *dict[ key ] = value*

Inclui um item ao dicionário.

In [ ]:
dados['DS5'] = 124549.07
'DS5' in dados

True

In [ ]:
dados

{'Jetta Variant': 88078.64,
 'Passat': 106161.94,
 'Crossfox': 72832.16,
 'DS5': 124549.07}

## *del dict[ key ]*

Remove o item de chave (*key*) do dicionário.

In [ ]:
dados

{'Jetta Variant': 88078.64,
 'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Passat': 106161.94}

In [ ]:
del dados['Passat']
dados

{'Jetta Variant': 88078.64, 'Crossfox': 72832.16, 'DS5': 124549.07}

# ATIVIDADE

## 1) Utilize o dicionário dados para responder a questão:

In [ ]:
dados = {
    'Passat': {
        'ano': 2012,
        'km': 50000,
        'valor': 75000,
        'acessorios': ['Airbag', 'ABS']
    },
    'Crossfox': {
        'ano': 2015,
        'km': 35000,
        'valor': 25000
    }
}

- 1) Testar se a chave acessorios existe no dicionário de informações do carro Crossfox (Resposta: False)

- 2) Testar se a chave acessorios existe no dicionário de informações do carro Passat (Resposta: True)

- 3) Obter o valor do carro Crossfox (Resposta: 25000)

- 4) Acessar o último acessório do carro Passat (Resposta: 'ABS')

In [ ]:
'acessorios' in dados['Crossfox']

False

In [ ]:
'acessorios' in dados['Passat']

True

In [ ]:
dados['Crossfox']['valor']

25000

In [ ]:
dados['Passat']['acessorios'][-1]

'ABS'

# 3.3 Métodos de dicionários

## *dict.update()*

Atualiza o dicionário.

In [ ]:
dados

{'Crossfox': 72832.16, 'DS5': 124549.07, 'Jetta Variant': 88078.64}

In [ ]:
dados.update({'Passat': 106161.94})    # atualiza dicionario incluindo esse par chave-valor se ele não estiver no dicionario
dados

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta Variant': 88078.64,
 'Passat': 106161.94}

In [ ]:
dados.update({'Passat': 106161.95, 'Fusca': 150000})  # se o par chave-valor já existir no dicionário você pode também atualizar o valor
dados

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta Variant': 88078.64,
 'Passat': 106161.95,
 'Fusca': 150000}

## *dict.copy()*

Cria uma cópia do dicionário.

In [ ]:
dadosCopy = dados.copy()

In [ ]:
dadosCopy

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta Variant': 88078.64,
 'Passat': 106161.95,
 'Fusca': 150000}

In [ ]:
del dadosCopy['Fusca']
dadosCopy

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta Variant': 88078.64,
 'Passat': 106161.95}

In [ ]:
dados

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta Variant': 88078.64,
 'Passat': 106161.95,
 'Fusca': 150000}

## *dict.pop(key[, default ])*

Se a chave for encontrada no dicionário, o item é removido e seu valor é retornado. Caso contrário, o valor especificado como *default* é retornado. Se o valor *default* não for fornecido e a chave não for encontrada no dicionário um erro será gerado.

In [ ]:
dadosCopy

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta Variant': 88078.64,
 'Passat': 106161.95}

In [ ]:
dadosCopy.pop('Passat')      # se a chave for encontrada no dicionário a função retorna o valor correspondente a essa chave e exclui o par chave-valor

106161.95

In [ ]:
dadosCopy

{'Crossfox': 72832.16, 'DS5': 124549.07, 'Jetta Variant': 88078.64}

In [ ]:
# dadosCopy.pop('Passat')

In [ ]:
dadosCopy.pop('Passat', 'Chave não encontrada')
#          dict.pop('key',  'default_value')  default_value: esse parâmetro nos permite fazer um tratamento de erro caso a chave não for encontrada no dicionário

'Chave não encontrada'

In [ ]:
dadosCopy.pop('DS5', 'Chave não encontrada')

124549.07

In [ ]:
dadosCopy

{'Crossfox': 72832.16, 'Jetta Variant': 88078.64}

## *dict.clear()*

Remove todos os itens do dicionário.

In [ ]:
dadosCopy.clear()

In [ ]:
dadosCopy

{}

# 3.4 Iterando em dicionários

In [ ]:
dados = {'Crossfox': 72832.16, 'DS5': 124549.07,  'Fusca': 150000,  'Jetta Variant': 88078.64,  'Passat': 106161.95}
dados

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Fusca': 150000,
 'Jetta Variant': 88078.64,
 'Passat': 106161.95}

## *dict.keys()*

Retorna uma lista contendo as chaves (*keys*) do dicionário.

In [ ]:
dados.keys()

dict_keys(['Crossfox', 'DS5', 'Fusca', 'Jetta Variant', 'Passat'])

In [ ]:
for key in dados.keys():    # acessa todos os itens na lista criada na função .keys() e para cada item na lista libera a saída com seu nome
  print(key)

Crossfox
DS5
Fusca
Jetta Variant
Passat


In [ ]:
for key in dados.keys(): # acessa a lista de chaves ['Crossfox', 'DS5', 'Fusca', 'Jetta Variant', 'Passat'] e para cada nome de carro printa o seu respectivo valor em dados acessando pela chave/nome
  print(dados[key])

72832.16
124549.07
150000
88078.64
106161.95


## *dict.values()*

Retorna uma lista com todos os valores (*values*) do dicionário.

In [ ]:
dados.values() # retorna apenas os valores do dicionário

dict_values([72832.16, 124549.07, 150000, 88078.64, 106161.95])

## *dict.items()*

Retorna uma lista contendo uma tupla para cada par chave-valor (*key-value*) do dicionário.

In [ ]:
dados.items()

dict_items([('Crossfox', 72832.16), ('DS5', 124549.07), ('Fusca', 150000), ('Jetta Variant', 88078.64), ('Passat', 106161.95)])

In [ ]:
for item in dados.items():
  print(item)

('Crossfox', 72832.16)
('DS5', 124549.07)
('Fusca', 150000)
('Jetta Variant', 88078.64)
('Passat', 106161.95)


In [ ]:
for key, value in dados.items():   # pegando as tuplas formadas pelos pares no dicionário e fazendo desempacotamento de tuplas
  print(key, value)

Crossfox 72832.16
DS5 124549.07
Fusca 150000
Jetta Variant 88078.64
Passat 106161.95


In [ ]:
for key, value in dados.items():
  if(value > 100000):
    print( key )

DS5
Fusca
Passat


# ATIVIDADE

## 1) Utilize o dicionário abaixo para responder esta atividade:

In [ ]:
dados = {
    'Crossfox': {'valor': 72000, 'ano': 2005},
    'DS5': {'valor': 125000, 'ano': 2015},
    'Fusca': {'valor': 150000, 'ano': 1976},
    'Jetta': {'valor': 88000, 'ano': 2010},
    'Passat': {'valor': 106000, 'ano': 1998}
}

 Crie o código que imprima somente os nomes dos veículos com ano de fabricação maior ou igual a 2000.

In [ ]:
for key, value in dados.items():
  if(value['ano'] >= 2000):
    print(key)

Crossfox
DS5
Jetta


# <font color=green> 4. FUNÇÕES E PACOTES
---
    
Funções são unidades de código reutilizáveis que realizam uma tarefa específica, podem receber alguma entrada e também podem retornar alguma resultado.

# 4.1 Built-in function

A linguagem Python possui várias funções integradas que estão sempre acessíveis. Algumas já utilizamos em nosso treinamento: type(), print(), zip(), len(), set() etc.
Essas Built-in functions são muito importantes pois podem nos ajudar a resolver problemas com menos código e de forma mais prática.

https://docs.python.org/3.6/library/functions.html

In [ ]:
dados = {'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}
dados

{'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}

In [ ]:
valores = []

for valor in dados.values():
  valores.append(valor)

valores

[88078.64, 106161.94, 72832.16]

In [ ]:
soma = 0

for valor in dados.values():
  soma += valor

soma

267072.74

In [ ]:
list(dados.values())

[88078.64, 106161.94, 72832.16]

In [ ]:
sum(dados.values())

267072.74

In [ ]:
# na built-in function sum é possível definir um parâmetro que diz o valor inicial do somatório
sum(dados.values(), start = 100000)

367072.74

In [ ]:
help(sum)

Help on built-in function sum in module builtins:

sum(iterable, /, start=0)
    Return the sum of a 'start' value (default: 0) plus an iterable of numbers
    
    When the iterable is empty, return the start value.
    This function is intended specifically for use with numeric values and may
    reject non-numeric types.



# 4.2 Definindo funções sem e com parâmetros

### Funções sem parâmetros

#### Formato padrão

```
def <nome>():
    <instruções>
```

In [ ]:
def media():
  valor = (1 + 2 + 3) / 3
  print(valor)

In [ ]:
media()

2.0


### Funções com parâmetros

#### Formato padrão

```
def <nome>(<param_1>, <param_2>, ..., <param_n>):
    <instruções>
```

In [ ]:
def media(num_1,num_2,num_3):
  valor = (num_1 + num_2 + num_3) / 3
  print(valor)

In [ ]:
media(2,3,10)

5.0


In [ ]:
media(24,45,66)

45.0


In [ ]:
def media(lista):
  valor = sum(lista) / len(lista)
  print(valor)

In [ ]:
resultado = media([1,2,3,4,5,6,7,8,9])

5.0


In [ ]:
resultado   # para armazenar certos valores de funções devemos definir um retorno para a função

In [ ]:
type(resultado)

NoneType

## ATIVIDADE

### 1) o cálculo da quilometragem média que um veículo rodou por ano. Não existe uma função pronta para o cálculo desta estatística em Python, e por isso precisamos escrever uma função que obtenha este valor.

Considere o conjunto de informações abaixo para responder o problema:

In [ ]:
dados = {
    'Crossfox': {'km': 35000, 'ano': 2005},
    'DS5': {'km': 17000, 'ano': 2015},
    'Fusca': {'km': 130000, 'ano': 1979},
    'Jetta': {'km': 56000, 'ano': 2011},
    'Passat': {'km': 62000, 'ano': 1999}
}

Considerando que o ano atual é '2019' faça uma função que receba como entrada os dados dos carros em formato de dicionário e o ano atual(2019)


Saída esperada:
- 2500.0
- 4250.0
- 3250.0
- 7000.0
- 3100.0

In [ ]:
def km_rodado (dict, ano_atual):
  for key, value in dados.items():
    km_total = value['km'] / (ano_atual - value['ano'])
    mensagem = (f'O carro {key} percorreu cerca de {km_total} Quilometros desde sua fabricação')
    print(mensagem)

km_rodado(dados, 2019)

O carro Crossfox percorreu cerca de 2500.0 Quilometros desde sua fabricação
O carro DS5 percorreu cerca de 4250.0 Quilometros desde sua fabricação
O carro Fusca percorreu cerca de 3250.0 Quilometros desde sua fabricação
O carro Jetta percorreu cerca de 7000.0 Quilometros desde sua fabricação
O carro Passat percorreu cerca de 3100.0 Quilometros desde sua fabricação


# 4.3 Definindo funções que retornam valores

### Funções que retornam um valor

#### Formato padrão

```
def <nome>(<param_1>, <param_2>, ..., <param_n>):
    <instruções>
    return <resultado>
```

In [ ]:
def media(lista):
  valor = sum(lista) / len(lista)
  return valor      # definindo um retorno, na chamada da função você tem uma saída no output e também pode usar-la para armazenar em alguma variável

In [ ]:
media([1,2,3,4,5])

3.0

In [ ]:
resultado = media([1,2,3,4,5,6,7])
resultado

4.0

In [ ]:
type(resultado)

float

### Funções que retornam mais de um valor

#### Formato padrão

```
def <nome>(<param_1>, <param_2>, ..., <param_n>):
    <instruções>
    return (<resultado_1>, <resultado_2>, ..., <resultado_n>)
```

In [ ]:
def media(lista):
  valor = sum(lista) / len(lista)
  return (valor, len(lista))

In [ ]:
media([1,2,3,4,5,6,7,8,9])

(5.0, 9)

In [ ]:
resultado, n = media([1,2,3,4,5,6,7,8,9])

In [ ]:
resultado

5.0

In [ ]:
n

9

In [ ]:
def km_media(dataset, ano_atual):
    result = {}
    for item in dataset.items():
        media = item[1]['km'] / (ano_atual - item[1]['ano'])
        result.update({ item[0]: media })
    return result

km_media(dados, 2019)

{'Crossfox': 2500.0,
 'DS5': 4250.0,
 'Fusca': 3250.0,
 'Jetta': 7000.0,
 'Passat': 3100.0}

# <font color=green> 5. PANDAS BÁSICO
---

**versão: 0.25.2**
  
Pandas é uma ferramenta de manipulação de dados de alto nível, construída com base no pacote Numpy. O pacote pandas possui estruturas de dados bastante interessantes para manipulação de dados e por isso é muito utilizado por cientistas de dados.


## Estruturas de Dados

### Series

Series são arrays unidimensionais rotulados capazes de armazenar qualquer tipo de dado. Os rótulos das linhas são chamados de **index**. A forma básica de criação de uma Series é a seguinte:


```
    s = pd.Series(dados, index = index)
```

O argumento *dados* pode ser um dicionário, uma lista, um array Numpy ou uma constante.

### DataFrames

DataFrame é uma estrutura de dados tabular bidimensional com rótulos nas linha e colunas. Como a Series, os DataFrames são capazes de armazenar qualquer tipo de dados.


```
    df = pd.DataFrame(dados, index = index, columns = columns)
```

O argumento *dados* pode ser um dicionário, uma lista, um array Numpy, uma Series e outro DataFrame.

**Documentação:** https://pandas.pydata.org/pandas-docs/version/0.25/

# 5.1 Estruturas de dados

In [ ]:
import pandas as pd

### Criando uma Series a partir de uma lista

In [ ]:
carros = ['Jetta Variant', 'Passat', 'Crossfox']
carros

['Jetta Variant', 'Passat', 'Crossfox']

Curiosidade: o Pandas trata strings como dados do tipo object.

In [ ]:
 # uma Serie é nada mais que um array rotulado e esse rotulo é o index dos elementos. criação de uma Serie -> pd.series(dados)
pd.Series(carros)

0    Jetta Variant
1           Passat
2         Crossfox
dtype: object

### Criando um DataFrame a partir de uma lista de dicionários

In [ ]:
dados = [
    {'Nome': 'Jetta Variant', 'Motor': 'Motor 4.0 Turbo', 'Ano': 2003, 'Quilometragem': 44410.0, 'Zero_km': False, 'Valor': 88078.64},
    {'Nome': 'Passat', 'Motor': 'Motor Diesel', 'Ano': 1991, 'Quilometragem': 5712.0, 'Zero_km': False, 'Valor': 106161.94},
    {'Nome': 'Crossfox', 'Motor': 'Motor Diesel V8', 'Ano': 1990, 'Quilometragem': 37123.0, 'Zero_km': False, 'Valor': 72832.16}
]

In [ ]:
dataset = pd.DataFrame(dados)

In [ ]:
dataset

Nome            Motor   Ano  Quilometragem  Zero_km      Valor
0  Jetta Variant  Motor 4.0 Turbo  2003        44410.0    False   88078.64
1         Passat     Motor Diesel  1991         5712.0    False  106161.94
2       Crossfox  Motor Diesel V8  1990        37123.0    False   72832.16

Alterando ordem das colunas

In [ ]:
dataset[['Nome', 'Motor','Ano', 'Quilometragem', 'Valor', 'Zero_km']]

Nome            Motor   Ano  Quilometragem      Valor  Zero_km
0  Jetta Variant  Motor 4.0 Turbo  2003        44410.0   88078.64    False
1         Passat     Motor Diesel  1991         5712.0  106161.94    False
2       Crossfox  Motor Diesel V8  1990        37123.0   72832.16    False

### Criando um DataFrame a partir de um dicionário

In [ ]:
dados = {
    'Nome': ['Jetta Variant', 'Passat', 'Crossfox'],
    'Motor': ['Motor 4.0 Turbo', 'Motor Diesel', 'Motor Diesel V8'],
    'Ano': [2003, 1991, 1990],
    'Quilometragem': [44410.0, 5712.0, 37123.0],
    'Zero_km': [False, False, False],
    'Valor': [88078.64, 106161.94, 72832.16]
}

In [ ]:
dataset = pd.DataFrame(dados)

In [ ]:
dataset

Nome            Motor   Ano  Quilometragem  Zero_km      Valor
0  Jetta Variant  Motor 4.0 Turbo  2003        44410.0    False   88078.64
1         Passat     Motor Diesel  1991         5712.0    False  106161.94
2       Crossfox  Motor Diesel V8  1990        37123.0    False   72832.16

### Criando um DataFrame a partir de uma arquivo externo

Link do arquivo externo utilizado no treinamento:

https://drive.google.com/file/d/1mF40OqATNmyX0Om5beEGzI6ScAmSkT-0/view?usp=sharing

In [ ]:
dataset = pd.read_csv('db.csv', sep = ';')

In [ ]:
dataset

Nome                  Motor   Ano  Quilometragem  \
0            Jetta Variant        Motor 4.0 Turbo  2003        44410.0   
1                   Passat           Motor Diesel  1991         5712.0   
2                 Crossfox        Motor Diesel V8  1990        37123.0   
3                      DS5        Motor 2.4 Turbo  2019            NaN   
4         Aston Martin DB4        Motor 2.4 Turbo  2006        25757.0   
..                     ...                    ...   ...            ...   
253           Phantom 2013               Motor V8  2014        27505.0   
254  Cadillac Ciel concept               Motor V8  1991        29981.0   
255             Classe GLK  Motor 5.0 V8 Bi-Turbo  2002        52637.0   
256       Aston Martin DB5           Motor Diesel  1996         7685.0   
257                  Macan        Motor Diesel V6  1992        50188.0   

     Zero_km                                         Acessórios      Valor  
0      False  ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
1      False  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
2      False  ['Piloto automático', 'Controle de estabilidad...   72832.16  
3       True  ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
4      False  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10  
..       ...                                                ...        ...  
253    False  ['Controle de estabilidade', 'Piloto automátic...   51759.58  
254    False  ['Bancos de couro', 'Painel digital', 'Sensor ...   51667.06  
255    False  ['Rodas de liga', 'Controle de tração', 'Câmbi...   68934.03  
256    False  ['Ar condicionado', '4 X 4', 'Câmbio automátic...  122110.90  
257    False  ['Central multimídia', 'Teto panorâmico', 'Vid...   90381.47  

[258 rows x 7 columns]

Mudando o rótulo do dataframe

In [ ]:
dataset = pd.read_csv('db.csv', sep = ';', index_col = 0)

In [ ]:
dataset

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                         
Jetta Variant                Motor 4.0 Turbo  2003        44410.0    False   
Passat                          Motor Diesel  1991         5712.0    False   
Crossfox                     Motor Diesel V8  1990        37123.0    False   
DS5                          Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4             Motor 2.4 Turbo  2006        25757.0    False   
...                                      ...   ...            ...      ...   
Phantom 2013                        Motor V8  2014        27505.0    False   
Cadillac Ciel concept               Motor V8  1991        29981.0    False   
Classe GLK             Motor 5.0 V8 Bi-Turbo  2002        52637.0    False   
Aston Martin DB5                Motor Diesel  1996         7685.0    False   
Macan                        Motor Diesel V6  1992        50188.0    False   

                                                              Acessórios  \
Nome                                                                       
Jetta Variant          ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox               ['Piloto automático', 'Controle de estabilidad...   
DS5                    ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4       ['Rodas de liga', '4 X 4', 'Central multimídia...   
...                                                                  ...   
Phantom 2013           ['Controle de estabilidade', 'Piloto automátic...   
Cadillac Ciel concept  ['Bancos de couro', 'Painel digital', 'Sensor ...   
Classe GLK             ['Rodas de liga', 'Controle de tração', 'Câmbi...   
Aston Martin DB5       ['Ar condicionado', '4 X 4', 'Câmbio automátic...   
Macan                  ['Central multimídia', 'Teto panorâmico', 'Vid...   

                           Valor  
Nome                              
Jetta Variant           88078.64  
Passat                 106161.94  
Crossfox                72832.16  
DS5                    124549.07  
Aston Martin DB4        92612.10  
...                          ...  
Phantom 2013            51759.58  
Cadillac Ciel concept   51667.06  
Classe GLK              68934.03  
Aston Martin DB5       122110.90  
Macan                   90381.47  

[258 rows x 6 columns]

In [ ]:
dados = {
    'Crossfox': {'km': 35000, 'ano': 2005},
    'DS5': {'km': 17000, 'ano': 2015},
    'Fusca': {'km': 130000, 'ano': 1979},
    'Jetta': {'km': 56000, 'ano': 2011},
    'Passat': {'km': 62000, 'ano': 1999}
}

In [ ]:
def km_media(dataset, ano_atual):
    result = {}
    for item in dataset.items():
        media = item[1]['km'] / (ano_atual - item[1]['ano'])
        item[1].update({ 'km_media': round(media, 2) })
        result.update({ item[0]: item[1] })

    return result

In [ ]:
dataset = pd.DataFrame(km_media(dados, 2023)).T
dataset

km     ano  km_media
Crossfox   35000.0  2005.0   1944.44
DS5        17000.0  2015.0   2125.00
Fusca     130000.0  1979.0   2954.55
Jetta      56000.0  2011.0   4666.67
Passat     62000.0  1999.0   2583.33

# 5.2 Seleções com DataFrames

In [ ]:
dataset = pd.read_csv('db.csv', sep = ';', index_col = 0)

In [ ]:
dataset.head()

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios      Valor  
Nome                                                                            
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox          ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10

### Selecionando colunas

In [ ]:
dataset['Valor'] # DataFrame[ ' Nome_rotulo/coluna ' ]   <- retorna apenas os valores da coluna escolhida juntamente com a coluna index

Nome
Jetta Variant             88078.64
Passat                   106161.94
Crossfox                  72832.16
DS5                      124549.07
Aston Martin DB4          92612.10
                           ...    
Phantom 2013              51759.58
Cadillac Ciel concept     51667.06
Classe GLK                68934.03
Aston Martin DB5         122110.90
Macan                     90381.47
Name: Valor, Length: 258, dtype: float64

In [ ]:
type(dataset['Valor'])

pandas.core.series.Series

Ou seja, ao selecionar valores de colunas em um DataFrame eu tenho como retorno uma Serie com o valor indíce e a coluna escolhida

In [ ]:
# Transformando a seleção de uma coluna em um DataFrame no formato DataFrame
dataset[['Valor']]

Valor
Nome                            
Jetta Variant           88078.64
Passat                 106161.94
Crossfox                72832.16
DS5                    124549.07
Aston Martin DB4        92612.10
...                          ...
Phantom 2013            51759.58
Cadillac Ciel concept   51667.06
Classe GLK              68934.03
Aston Martin DB5       122110.90
Macan                   90381.47

[258 rows x 1 columns]

In [ ]:
type(dataset[['Valor']])

pandas.core.frame.DataFrame

### Selecionando linhas - [ i : j ]

<font color=red>**Observação:**</font> A indexação tem origem no zero e nos fatiamentos (*slices*) a linha com índice i é **incluída** e a linha com índice j **não é incluída** no resultado.

In [ ]:
dataset[0:3]

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                           
Jetta Variant  Motor 4.0 Turbo  2003        44410.0    False   
Passat            Motor Diesel  1991         5712.0    False   
Crossfox       Motor Diesel V8  1990        37123.0    False   

                                                      Acessórios      Valor  
Nome                                                                         
Jetta Variant  ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat         ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox       ['Piloto automático', 'Controle de estabilidad...   72832.16

### Utilizando .loc para seleções

<font color=red>**Observação:**</font> Seleciona um grupo de linhas e colunas segundo os rótulos ou uma matriz booleana.

In [ ]:
dataset.loc['Passat'] # retorna uma Serie com informações da linha do Passat

Motor                                                 Motor Diesel
Ano                                                           1991
Quilometragem                                               5712.0
Zero_km                                                      False
Acessórios       ['Central multimídia', 'Teto panorâmico', 'Fre...
Valor                                                    106161.94
Name: Passat, dtype: object

In [ ]:
dataset.loc[['Passat', 'DS5']] # retorna informações de duas linhas

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                    
Passat     Motor Diesel  1991         5712.0    False   
DS5     Motor 2.4 Turbo  2019            NaN     True   

                                               Acessórios      Valor  
Nome                                                                  
Passat  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
DS5     ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07

In [ ]:
dataset.loc[['Passat', 'DS5'], ['Motor', 'Valor']]    # seleciona as informações das linhas Passat e DS5, porém, só retorna as informações presentes nas colunas motor e valor

Motor      Valor
Nome                              
Passat     Motor Diesel  106161.94
DS5     Motor 2.4 Turbo  124549.07

In [ ]:
dataset.loc[:, ['Motor', 'Valor']]  # selecionando as informações das colunas Motor e Valor em todas as linhas

Motor      Valor
Nome                                                   
Jetta Variant                Motor 4.0 Turbo   88078.64
Passat                          Motor Diesel  106161.94
Crossfox                     Motor Diesel V8   72832.16
DS5                          Motor 2.4 Turbo  124549.07
Aston Martin DB4             Motor 2.4 Turbo   92612.10
...                                      ...        ...
Phantom 2013                        Motor V8   51759.58
Cadillac Ciel concept               Motor V8   51667.06
Classe GLK             Motor 5.0 V8 Bi-Turbo   68934.03
Aston Martin DB5                Motor Diesel  122110.90
Macan                        Motor Diesel V6   90381.47

[258 rows x 2 columns]

### Utilizando .iloc para seleções

<font color=red>**Observação:**</font> Seleciona com base nos índices, ou seja, se baseia na posição das informações.

In [ ]:
dataset.head()

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios      Valor  
Nome                                                                            
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox          ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10

In [ ]:
dataset.iloc[1] #retorna informações do índice em formato de Serie 1 -> Passat

Motor                                                 Motor Diesel
Ano                                                           1991
Quilometragem                                               5712.0
Zero_km                                                      False
Acessórios       ['Central multimídia', 'Teto panorâmico', 'Fre...
Valor                                                    106161.94
Name: Passat, dtype: object

In [ ]:
dataset.iloc[[1]] #retorna as mesmas informações porém como um DataFrame

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                 
Passat  Motor Diesel  1991         5712.0    False   

                                               Acessórios      Valor  
Nome                                                                  
Passat  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94

In [ ]:
dataset.iloc[1:4]

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                      
Passat       Motor Diesel  1991         5712.0    False   
Crossfox  Motor Diesel V8  1990        37123.0    False   
DS5       Motor 2.4 Turbo  2019            NaN     True   

                                                 Acessórios      Valor  
Nome                                                                    
Passat    ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox  ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5       ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07

In [ ]:
dataset.iloc[1:4, [0, 5, 2]] # retorna as linhas de 1 até indice 3, porém, apenas com informações das colunas de indice 0,5 e 2

Motor      Valor  Quilometragem
Nome                                               
Passat       Motor Diesel  106161.94         5712.0
Crossfox  Motor Diesel V8   72832.16        37123.0
DS5       Motor 2.4 Turbo  124549.07            NaN

In [ ]:
dataset.iloc[[1, 42, 22], [0, 5, 2]]  # retorna informações das colunas 0, 5 e 2 referentes as linhas de indíce 1, 42 e 22

Motor      Valor  Quilometragem
Nome                                                          
Passat                  Motor Diesel  106161.94         5712.0
Optima                 Motor 1.8 16v   86641.34            NaN
Lamborghini Obvious  Motor Diesel V6  133529.84        98079.0

In [ ]:
dataset.iloc[:, [0, 5, 2]]

Motor      Valor  Quilometragem
Nome                                                                  
Jetta Variant                Motor 4.0 Turbo   88078.64        44410.0
Passat                          Motor Diesel  106161.94         5712.0
Crossfox                     Motor Diesel V8   72832.16        37123.0
DS5                          Motor 2.4 Turbo  124549.07            NaN
Aston Martin DB4             Motor 2.4 Turbo   92612.10        25757.0
...                                      ...        ...            ...
Phantom 2013                        Motor V8   51759.58        27505.0
Cadillac Ciel concept               Motor V8   51667.06        29981.0
Classe GLK             Motor 5.0 V8 Bi-Turbo   68934.03        52637.0
Aston Martin DB5                Motor Diesel  122110.90         7685.0
Macan                        Motor Diesel V6   90381.47        50188.0

[258 rows x 3 columns]

## Atividade

### 1) Considere os seguintes dados:

In [ ]:
dados = {
    'Nome': ['Jetta', 'Passat', 'Crossfox', 'DS5', 'Fusca'],
    'Motor': ['Motor 4.0 Turbo', 'Motor Diesel', 'Motor Diesel V8', 'Motor 2.0', 'Motor 1.6'],
    'Ano': [2019, 2003, 1991, 2019, 1990],
    'Quilometragem': [0.0, 5712.0, 37123.0, 0.0, 120000.0],
    'Zero_km': [True, False, False, True, False],
    'Valor': [88000.0, 106000.0, 72000.0, 89000.0, 32000.0]
}

In [ ]:
dataset = pd.DataFrame(dados)
dataset

Nome            Motor   Ano  Quilometragem  Zero_km     Valor
0     Jetta  Motor 4.0 Turbo  2019            0.0     True   88000.0
1    Passat     Motor Diesel  2003         5712.0    False  106000.0
2  Crossfox  Motor Diesel V8  1991        37123.0    False   72000.0
3       DS5        Motor 2.0  2019            0.0     True   89000.0
4     Fusca        Motor 1.6  1990       120000.0    False   32000.0


Crie três códigos que selecionam somente as informações de Nome, Ano, Quilometragem e Valor dos carros Passat e Crossfox.

In [ ]:
dataset[['Nome', 'Ano', 'Quilometragem', 'Valor']][1:3]  # selecionando usando a forma convencional

Nome   Ano  Quilometragem     Valor
1    Passat  2003         5712.0  106000.0
2  Crossfox  1991        37123.0   72000.0

In [ ]:
dataset.loc[[1, 2], ['Nome', 'Ano', 'Quilometragem', 'Valor']]  # usando .loc

Nome   Ano  Quilometragem     Valor
1    Passat  2003         5712.0  106000.0
2  Crossfox  1991        37123.0   72000.0

In [ ]:
dataset.iloc[[1, 2], [0, 2, 3, 5]]  # usando .iloc

Nome   Ano  Quilometragem     Valor
1    Passat  2003         5712.0  106000.0
2  Crossfox  1991        37123.0   72000.0

### 2) Utilize o código abaixo para resolver a questão:

In [ ]:
import pandas as pd

dados = {
    'Motor': ['Motor 4.0 Turbo', 'Motor Diesel', 'Motor Diesel V8', 'Motor 2.0', 'Motor 1.6'],
    'Ano': [2019, 2003, 1991, 2019, 1990],
    'Quilometragem': [0.0, 5712.0, 37123.0, 0.0, 120000.0],
    'Zero_km': [True, False, False, True, False],
    'Valor': [88000.0, 106000.0, 72000.0, 89000.0, 32000.0]
}

dataset = pd.DataFrame(dados, index = ['Jetta', 'Passat', 'Crossfox', 'DS5', 'Fusca'])

Note que utilizamos os nomes dos veículos como índice do DataFrame.

Escreva o código que têm como saída o seguinte DataFrame:


                     Motor  	          Valor
    Passat	Motor Diesel	106000.0
    DS5	     Motor 2.0	     89000.0


In [ ]:
dataset.loc[['Passat', 'DS5'], ['Motor', 'Valor']]  # usando o .loc

Motor     Valor
Passat  Motor Diesel  106000.0
DS5        Motor 2.0   89000.0

In [ ]:
dataset.iloc[[1,3], [0,-1]] # usando o .iloc

Motor     Valor
Passat  Motor Diesel  106000.0
DS5        Motor 2.0   89000.0

# 5.3 Queries com DataFrames

In [ ]:
dataset.head()

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios      Valor  
Nome                                                                            
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox          ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10

In [ ]:
dataset.Motor # retorna uma Serie com a coluna index e a coluna motor + informações da coluna motor

Nome
Jetta Variant                  Motor 4.0 Turbo
Passat                            Motor Diesel
Crossfox                       Motor Diesel V8
DS5                            Motor 2.4 Turbo
Aston Martin DB4               Motor 2.4 Turbo
                                 ...          
Phantom 2013                          Motor V8
Cadillac Ciel concept                 Motor V8
Classe GLK               Motor 5.0 V8 Bi-Turbo
Aston Martin DB5                  Motor Diesel
Macan                          Motor Diesel V6
Name: Motor, Length: 258, dtype: object

In [ ]:
# fazendo uma querie(consulta) que retorna uma Serie dizendo se os carros tem Motor Diesel ou não -> Serie booleana
select = dataset.Motor == 'Motor Diesel'

In [ ]:
type(select)

pandas.core.series.Series

In [ ]:
dataset[select] # retorna do Dataset apenas os carros contidos como True na Serie/matriz booleana do select, ou seja, apenas carros com Motor Diesel

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                
Passat                 Motor Diesel  1991         5712.0    False   
Effa Hafei Picape Baú  Motor Diesel  1991       102959.0    False   
Sorento                Motor Diesel  2019            NaN     True   
New Fiesta Hatch       Motor Diesel  2017       118895.0    False   
Kangoo Express         Motor Diesel  2007        29132.0    False   
Fit                    Motor Diesel  2013        44329.0    False   
Cielo Hatch            Motor Diesel  2019            NaN     True   
Symbol                 Motor Diesel  2016       117714.0    False   
A4 Sedan               Motor Diesel  2002        30511.0    False   
A4 Avant               Motor Diesel  2014        17357.0    False   
Silver Shadow          Motor Diesel  2015        99052.0    False   
Camry                  Motor Diesel  2019            NaN     True   
Clio                   Motor Diesel  1990        66437.0    False   
Série 3 M3 Coupé       Motor Diesel  2013        19896.0    False   
Gallardo LP 560 – 4    Motor Diesel  2000       113045.0    False   
Chana Utility          Motor Diesel  2009        52034.0    False   
TT Roadster            Motor Diesel  1998       107392.0    False   
Aston Martin Virage    Motor Diesel  2019            NaN     True   
Sandero                Motor Diesel  2015        34783.0    False   
XKR                    Motor Diesel  1996        85127.0    False   
X5                     Motor Diesel  2002        13606.0    False   
Dodge Dakota           Motor Diesel  1993        71544.0    False   
i30 CW                 Motor Diesel  2015         8497.0    False   
Série 7 Sedã           Motor Diesel  2019            NaN     True   
V60                    Motor Diesel  2004        91840.0    False   
Aston Martin DB5       Motor Diesel  1996         7685.0    False   

                                                              Acessórios  \
Nome                                                                       
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Effa Hafei Picape Baú  ['Controle de estabilidade', 'Painel digital',...   
Sorento                ['Sensor de chuva', 'Câmera de estacionamento'...   
New Fiesta Hatch       ['Sensor de estacionamento', 'Travas elétricas...   
Kangoo Express         ['Bancos de couro', 'Câmbio automático', 'Pilo...   
Fit                    ['Freios ABS', 'Câmera de estacionamento', 'Câ...   
Cielo Hatch            ['Painel digital', 'Central multimídia', 'Câme...   
Symbol                 ['4 X 4', 'Piloto automático', 'Sensor crepusc...   
A4 Sedan               ['Câmera de estacionamento', '4 X 4', 'Travas ...   
A4 Avant               ['Teto panorâmico', '4 X 4', 'Bancos de couro'...   
Silver Shadow          ['4 X 4', 'Central multimídia', 'Ar condiciona...   
Camry                  ['Travas elétricas', 'Rodas de liga', 'Sensor ...   
Clio                   ['4 X 4', 'Câmbio automático', 'Freios ABS', '...   
Série 3 M3 Coupé       ['Bancos de couro', 'Sensor de chuva', 'Ar con...   
Gallardo LP 560 – 4    ['Freios ABS', 'Teto panorâmico', 'Central mul...   
Chana Utility          ['Ar condicionado', 'Sensor de estacionamento'...   
TT Roadster            ['Ar condicionado', 'Sensor de estacionamento'...   
Aston Martin Virage    ['Travas elétricas', 'Controle de tração', 'Câ...   
Sandero                ['Piloto automático', 'Teto panorâmico', 'Câmb...   
XKR                    ['Controle de tração', 'Ar condicionado', 'Tra...   
X5                     ['Controle de tração', 'Piloto automático', 'T...   
Dodge Dakota           ['Controle de tração', 'Sensor de estacionamen...   
i30 CW                 ['Sensor de chuva', 'Vidros elétricos', 'Contr...   
Série 7 Sedã           ['Vidros elétricos', 'Travas elétricas', 'Roda...   
V60                    ['Rodas de liga', 'Sensor crepuscular', 'Ar co...   
Aston Martin DB5       ['Ar condicionado', '4

In [ ]:
(dataset.Motor == 'Motor Diesel') & (dataset.Zero_km == True) # retorna uma Serie booleana que mostra se os carros atendem ou não essa condição

Nome
Jetta Variant            False
Passat                   False
Crossfox                 False
DS5                      False
Aston Martin DB4         False
                         ...  
Phantom 2013             False
Cadillac Ciel concept    False
Classe GLK               False
Aston Martin DB5         False
Macan                    False
Length: 258, dtype: bool

In [ ]:
dataset[ (dataset.Motor == 'Motor Diesel') & (dataset.Zero_km == True) ]    # usando Matriz booleanas retorna um DataFrame exibindo os carros que atendem a condição

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Sorento              Motor Diesel  2019            NaN     True   
Cielo Hatch          Motor Diesel  2019            NaN     True   
Camry                Motor Diesel  2019            NaN     True   
Aston Martin Virage  Motor Diesel  2019            NaN     True   
Série 7 Sedã         Motor Diesel  2019            NaN     True   

                                                            Acessórios  \
Nome                                                                     
Sorento              ['Sensor de chuva', 'Câmera de estacionamento'...   
Cielo Hatch          ['Painel digital', 'Central multimídia', 'Câme...   
Camry                ['Travas elétricas', 'Rodas de liga', 'Sensor ...   
Aston Martin Virage  ['Travas elétricas', 'Controle de tração', 'Câ...   
Série 7 Sedã         ['Vidros elétricos', 'Travas elétricas', 'Roda...   

                         Valor  
Nome                            
Sorento               81399.35  
Cielo Hatch          145197.70  
Camry                138597.27  
Aston Martin Virage   97290.18  
Série 7 Sedã          67539.79

### Utilizando o método query

dataset.query(' Condição ')

In [ ]:
dataset.query(' Motor == "Motor Diesel" and Zero_km == True ') # retorna Data Frame com carros com Motor Diesel e Zero km

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Sorento              Motor Diesel  2019            NaN     True   
Cielo Hatch          Motor Diesel  2019            NaN     True   
Camry                Motor Diesel  2019            NaN     True   
Aston Martin Virage  Motor Diesel  2019            NaN     True   
Série 7 Sedã         Motor Diesel  2019            NaN     True   

                                                            Acessórios  \
Nome                                                                     
Sorento              ['Sensor de chuva', 'Câmera de estacionamento'...   
Cielo Hatch          ['Painel digital', 'Central multimídia', 'Câme...   
Camry                ['Travas elétricas', 'Rodas de liga', 'Sensor ...   
Aston Martin Virage  ['Travas elétricas', 'Controle de tração', 'Câ...   
Série 7 Sedã         ['Vidros elétricos', 'Travas elétricas', 'Roda...   

                         Valor  
Nome                            
Sorento               81399.35  
Cielo Hatch          145197.70  
Camry                138597.27  
Aston Martin Virage   97290.18  
Série 7 Sedã          67539.79

## Atividade

### 1) Utilize o seguinte DataFrame para responder a questão:

In [ ]:
dados = {
    'Motor': ['Motor 4.0 Turbo', 'Motor Diesel', 'Motor Diesel V8', 'Motor Diesel', 'Motor 1.6'],
    'Ano': [2019, 2003, 1991, 2019, 1990],
    'Quilometragem': [0.0, 5712.0, 37123.0, 0.0, 120000.0],
    'Zero_km': [True, False, False, True, False],
    'Valor': [88000.0, 106000.0, 72000.0, 89000.0, 32000.0]
}

dataset = pd.DataFrame(dados, index = ['Jetta', 'Passat', 'Crossfox', 'DS5', 'Fusca'])


                     Motor	              Ano	Quilometragem Zero_km	   Valor
    Jetta	   Motor 4.0 Turbo  2019	 0.0	       	    True	 	      88000.0
    Passat	Motor Diesel	   2003	 5712.0		    False		    106000.0
    DS5	    Motor Diesel	    2019	0.0	    	       True	 	      89000.0


Crie o código que retorne esse dataframe acima:

In [ ]:
select = (dataset.Motor == 'Motor Diesel') | (dataset.Zero_km == True)
dataset[select]

Motor   Ano  Quilometragem  Zero_km     Valor
Jetta   Motor 4.0 Turbo  2019            0.0     True   88000.0
Passat     Motor Diesel  2003         5712.0    False  106000.0
DS5        Motor Diesel  2019            0.0     True   89000.0

In [ ]:
dataset.query(' (Motor == "Motor Diesel") | (Zero_km == True) ')

Motor   Ano  Quilometragem  Zero_km     Valor
Jetta   Motor 4.0 Turbo  2019            0.0     True   88000.0
Passat     Motor Diesel  2003         5712.0    False  106000.0
DS5        Motor Diesel  2019            0.0     True   89000.0

# 5.4 Iterando com DataFrames

In [ ]:
dataset.head()

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios      Valor  
Nome                                                                            
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox          ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10

In [ ]:
for item in dataset:
    print(item)

Motor
Ano
Quilometragem
Zero_km
Acessórios
Valor


Porém, isso não tem muita utilidade para nós. O Pandas nos disponibiliza uma ferramenta interessante chamada iterrows(), que nos ajuda a criar um iterador

O resultado disso será uma lista contendo diversas tuplas, cada uma composta pelo índice do nosso dataframe (os nomes dos veículos) e uma series com as informações da linha correspondente a esse índice. Isso nos permitirá acessar tais informações.

In [ ]:
list(dataset.iterrows()) # itera o dataset e retorna para cada elemento uma tupla contendo seu index e uma serie com as colunas e valores respectivos

[('Jetta Variant',
  Motor                                              Motor 4.0 Turbo
  Ano                                                           2003
  Quilometragem                                              44410.0
  Zero_km                                                      False
  Acessórios       ['Rodas de liga', 'Travas elétricas', 'Piloto ...
  Valor                                                     88078.64
  Name: Jetta Variant, dtype: object),
 ('Passat',
  Motor                                                 Motor Diesel
  Ano                                                           1991
  Quilometragem                                               5712.0
  Zero_km                                                      False
  Acessórios       ['Central multimídia', 'Teto panorâmico', 'Fre...
  Valor                                                    106161.94
  Name: Passat, dtype: object),
 ('Crossfox',
  Motor                                              Mot

In [ ]:
for index, row in dataset.iterrows():
  if(2019 - row['Ano'] != 0): # se o carro não é de 2019
    dataset.loc[index, 'Km_media'] = row['Quilometragem'] / (2019 - row['Ano']) # atualiza o DataFrame com a coluna Km_media que é uma divisão entre a coluna Km e (2019 - coluna(ano)) -> KM / (2019 - ano)
  else: # se não, ou seja, é de 2019
    dataset.loc[index, 'Km_media'] = 0 # Atualiza a coluna km_media dessa linha como 0

dataset

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                         
Jetta Variant                Motor 4.0 Turbo  2003        44410.0    False   
Passat                          Motor Diesel  1991         5712.0    False   
Crossfox                     Motor Diesel V8  1990        37123.0    False   
DS5                          Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4             Motor 2.4 Turbo  2006        25757.0    False   
...                                      ...   ...            ...      ...   
Phantom 2013                        Motor V8  2014        27505.0    False   
Cadillac Ciel concept               Motor V8  1991        29981.0    False   
Classe GLK             Motor 5.0 V8 Bi-Turbo  2002        52637.0    False   
Aston Martin DB5                Motor Diesel  1996         7685.0    False   
Macan                        Motor Diesel V6  1992        50188.0    False   

                                                              Acessórios  \
Nome                                                                       
Jetta Variant          ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox               ['Piloto automático', 'Controle de estabilidad...   
DS5                    ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4       ['Rodas de liga', '4 X 4', 'Central multimídia...   
...                                                                  ...   
Phantom 2013           ['Controle de estabilidade', 'Piloto automátic...   
Cadillac Ciel concept  ['Bancos de couro', 'Painel digital', 'Sensor ...   
Classe GLK             ['Rodas de liga', 'Controle de tração', 'Câmbi...   
Aston Martin DB5       ['Ar condicionado', '4 X 4', 'Câmbio automátic...   
Macan                  ['Central multimídia', 'Teto panorâmico', 'Vid...   

                           Valor     Km_media  
Nome                                           
Jetta Variant           88078.64  2775.625000  
Passat                 106161.94   204.000000  
Crossfox                72832.16  1280.103448  
DS5                    124549.07     0.000000  
Aston Martin DB4        92612.10  1981.307692  
...                          ...          ...  
Phantom 2013            51759.58  5501.000000  
Cadillac Ciel concept   51667.06  1070.750000  
Classe GLK              68934.03  3096.294118  
Aston Martin DB5       122110.90   334.130435  
Macan                   90381.47  1858.814815  

[258 rows x 7 columns]

In [ ]:
for index, row in dataset.iterrows():
  if row['Valor'] < 55001:
    dataset.loc[index, 'Critica_preco'] = 'Em conta'
  else:
    dataset.loc[index, 'Critica_preco'] = 'caro'

dataset

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                         
Jetta Variant                Motor 4.0 Turbo  2003        44410.0    False   
Passat                          Motor Diesel  1991         5712.0    False   
Crossfox                     Motor Diesel V8  1990        37123.0    False   
DS5                          Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4             Motor 2.4 Turbo  2006        25757.0    False   
...                                      ...   ...            ...      ...   
Phantom 2013                        Motor V8  2014        27505.0    False   
Cadillac Ciel concept               Motor V8  1991        29981.0    False   
Classe GLK             Motor 5.0 V8 Bi-Turbo  2002        52637.0    False   
Aston Martin DB5                Motor Diesel  1996         7685.0    False   
Macan                        Motor Diesel V6  1992        50188.0    False   

                                                              Acessórios  \
Nome                                                                       
Jetta Variant          ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox               ['Piloto automático', 'Controle de estabilidad...   
DS5                    ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4       ['Rodas de liga', '4 X 4', 'Central multimídia...   
...                                                                  ...   
Phantom 2013           ['Controle de estabilidade', 'Piloto automátic...   
Cadillac Ciel concept  ['Bancos de couro', 'Painel digital', 'Sensor ...   
Classe GLK             ['Rodas de liga', 'Controle de tração', 'Câmbi...   
Aston Martin DB5       ['Ar condicionado', '4 X 4', 'Câmbio automátic...   
Macan                  ['Central multimídia', 'Teto panorâmico', 'Vid...   

                           Valor     Km_media Critica_preco  
Nome                                                         
Jetta Variant           88078.64  2775.625000          caro  
Passat                 106161.94   204.000000          caro  
Crossfox                72832.16  1280.103448          caro  
DS5                    124549.07     0.000000          caro  
Aston Martin DB4        92612.10  1981.307692          caro  
...                          ...          ...           ...  
Phantom 2013            51759.58  5501.000000      Em conta  
Cadillac Ciel concept   51667.06  1070.750000      Em conta  
Classe GLK              68934.03  3096.294118          caro  
Aston Martin DB5       122110.90   334.130435          caro  
Macan                   90381.47  1858.814815          caro  

[258 rows x 8 columns]

In [ ]:
list(dataset.items()) # retorna para cada coluna os seus valores e informações

[('Motor',
  Nome
  Jetta Variant                  Motor 4.0 Turbo
  Passat                            Motor Diesel
  Crossfox                       Motor Diesel V8
  DS5                            Motor 2.4 Turbo
  Aston Martin DB4               Motor 2.4 Turbo
                                   ...          
  Phantom 2013                          Motor V8
  Cadillac Ciel concept                 Motor V8
  Classe GLK               Motor 5.0 V8 Bi-Turbo
  Aston Martin DB5                  Motor Diesel
  Macan                          Motor Diesel V6
  Name: Motor, Length: 258, dtype: object),
 ('Ano',
  Nome
  Jetta Variant            2003
  Passat                   1991
  Crossfox                 1990
  DS5                      2019
  Aston Martin DB4         2006
                           ... 
  Phantom 2013             2014
  Cadillac Ciel concept    1991
  Classe GLK               2002
  Aston Martin DB5         1996
  Macan                    1992
  Name: Ano, Length: 258, dtype

In [ ]:
list(dataset.itertuples()) # itera o dataset retornando cada um dos elementos como uma tupla

[Pandas(Index='Jetta Variant', Motor='Motor 4.0 Turbo', Ano=2003, Quilometragem=44410.0, Zero_km=False, Acessórios="['Rodas de liga', 'Travas elétricas', 'Piloto automático', 'Bancos de couro', 'Ar condicionado', 'Sensor de estacionamento', 'Sensor crepuscular', 'Sensor de chuva']", Valor=88078.64, Km_media=2775.625, Critica_preco='caro'),
 Pandas(Index='Passat', Motor='Motor Diesel', Ano=1991, Quilometragem=5712.0, Zero_km=False, Acessórios="['Central multimídia', 'Teto panorâmico', 'Freios ABS', '4 X 4', 'Painel digital', 'Piloto automático', 'Bancos de couro', 'Câmera de estacionamento']", Valor=106161.94, Km_media=204.0, Critica_preco='caro'),
 Pandas(Index='Crossfox', Motor='Motor Diesel V8', Ano=1990, Quilometragem=37123.0, Zero_km=False, Acessórios="['Piloto automático', 'Controle de estabilidade', 'Sensor crepuscular', 'Freios ABS', 'Câmbio automático', 'Bancos de couro', 'Central multimídia', 'Vidros elétricos']", Valor=72832.16, Km_media=1280.103448275862, Critica_preco='caro

# 5.5 Tratamento de dados

In [ ]:
dataset.head()

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios  \
Nome                                                                  
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox          ['Piloto automático', 'Controle de estabilidad...   
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   

                      Valor     Km_media Critica_preco  
Nome                                                    
Jetta Variant      88078.64  2775.625000          caro  
Passat            106161.94   204.000000          caro  
Crossfox           72832.16  1280.103448          caro  
DS5               124549.07     0.000000          caro  
Aston Martin DB4   92612.10  1981.307692          caro

In [ ]:
dataset.info()   # retorna informações do dataset

<class 'pandas.core.frame.DataFrame'>
Index: 258 entries, Jetta Variant to Macan
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Motor          258 non-null    object 
 1   Ano            258 non-null    int64  
 2   Quilometragem  197 non-null    float64
 3   Zero_km        258 non-null    bool   
 4   Acessórios     258 non-null    object 
 5   Valor          258 non-null    float64
 6   Km_media       258 non-null    float64
 7   Critica_preco  258 non-null    object 
dtypes: bool(1), float64(3), int64(1), object(3)
memory usage: 24.5+ KB


Note que na coluna Quilometragem existem 197 registros não nulos, ou seja, dentre as linhas existem 61 linhas que possuem o valor nessa coluna como nulo. Conseguimos com tratamento de dados corrigir isso


In [ ]:
# o Pandas nos permite selecionar informações que são nulas ou não
# dataset.Variavel.isna()
dataset.Quilometragem.isna()    # retorna uma Series booleana onde quando o valor for Nulo mostra como True

Nome
Jetta Variant            False
Passat                   False
Crossfox                 False
DS5                       True
Aston Martin DB4         False
                         ...  
Phantom 2013             False
Cadillac Ciel concept    False
Classe GLK               False
Aston Martin DB5         False
Macan                    False
Name: Quilometragem, Length: 258, dtype: bool

In [ ]:
dataset[dataset.Quilometragem.isna()] # retorna um dataframe apenas com os carros que tem o valor Quilometragem como nulo

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                  
DS5                   Motor 2.4 Turbo  2019            NaN     True   
A5                    Motor 4.0 Turbo  2019            NaN     True   
J5                           Motor V6  2019            NaN     True   
A3                       Motor 1.0 8v  2019            NaN     True   
Série 1 M                    Motor V8  2019            NaN     True   
...                               ...   ...            ...      ...   
Lamborghini Reventón  Motor 4.0 Turbo  2019            NaN     True   
Benni Mini                   Motor V8  2019            NaN     True   
Uno                   Motor Diesel V6  2019            NaN     True   
Santa Fe                Motor 3.0 32v  2019            NaN     True   
XC60                  Motor 4.0 Turbo  2019            NaN     True   

                                                             Acessórios  \
Nome                                                                      
DS5                   ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
A5                    ['Câmbio automático', 'Câmera de estacionament...   
J5                    ['Sensor crepuscular', 'Painel digital', 'Roda...   
A3                    ['4 X 4', 'Piloto automático', 'Central multim...   
Série 1 M             ['Controle de estabilidade', 'Central multimíd...   
...                                                                 ...   
Lamborghini Reventón  ['Controle de tração', 'Ar condicionado', 'Cen...   
Benni Mini            ['Sensor crepuscular', 'Câmbio automático', 'C...   
Uno                   ['Central multimídia', 'Sensor crepuscular', '...   
Santa Fe              ['Travas elétricas', 'Ar condicionado', '4 X 4...   
XC60                  ['Painel digital', 'Piloto automático', 'Centr...   

                          Valor  Km_media Critica_preco  
Nome                                                     
DS5                   124549.07       0.0          caro  
A5                     56445.20       0.0          caro  
J5                     53183.38       0.0      Em conta  
A3                     88552.39       0.0          caro  
Série 1 M              94564.40       0.0          caro  
...                         ...       ...           ...  
Lamborghini Reventón   67664.86       0.0          caro  
Benni Mini            126247.84       0.0          caro  
Uno                   128852.21       0.0          caro  
Santa Fe              129415.33       0.0          caro  
XC60                   77675.79       0.0          caro  

[61 rows x 8 columns]

In [ ]:
dataset.fillna(0, inplace = True)
# preenche valores nulos -> dataset.fillna(valor_desejado, inplace = True) <- o parâmetro que define a variavel inplace serve para definir se a alteração é feita no dataset e não só na visualização

In [ ]:
dataset.query('Zero_km == True')

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                  
DS5                   Motor 2.4 Turbo  2019            0.0     True   
A5                    Motor 4.0 Turbo  2019            0.0     True   
J5                           Motor V6  2019            0.0     True   
A3                       Motor 1.0 8v  2019            0.0     True   
Série 1 M                    Motor V8  2019            0.0     True   
...                               ...   ...            ...      ...   
Lamborghini Reventón  Motor 4.0 Turbo  2019            0.0     True   
Benni Mini                   Motor V8  2019            0.0     True   
Uno                   Motor Diesel V6  2019            0.0     True   
Santa Fe                Motor 3.0 32v  2019            0.0     True   
XC60                  Motor 4.0 Turbo  2019            0.0     True   

                                                             Acessórios  \
Nome                                                                      
DS5                   ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
A5                    ['Câmbio automático', 'Câmera de estacionament...   
J5                    ['Sensor crepuscular', 'Painel digital', 'Roda...   
A3                    ['4 X 4', 'Piloto automático', 'Central multim...   
Série 1 M             ['Controle de estabilidade', 'Central multimíd...   
...                                                                 ...   
Lamborghini Reventón  ['Controle de tração', 'Ar condicionado', 'Cen...   
Benni Mini            ['Sensor crepuscular', 'Câmbio automático', 'C...   
Uno                   ['Central multimídia', 'Sensor crepuscular', '...   
Santa Fe              ['Travas elétricas', 'Ar condicionado', '4 X 4...   
XC60                  ['Painel digital', 'Piloto automático', 'Centr...   

                          Valor  Km_media Critica_preco  
Nome                                                     
DS5                   124549.07       0.0          caro  
A5                     56445.20       0.0          caro  
J5                     53183.38       0.0      Em conta  
A3                     88552.39       0.0          caro  
Série 1 M              94564.40       0.0          caro  
...                         ...       ...           ...  
Lamborghini Reventón   67664.86       0.0          caro  
Benni Mini            126247.84       0.0          caro  
Uno                   128852.21       0.0          caro  
Santa Fe              129415.33       0.0          caro  
XC60                   77675.79       0.0          caro  

[61 rows x 8 columns]

In [ ]:
dataset = pd.read_csv('db.csv', sep = ";", index_col = 0)

In [ ]:
dataset

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                         
Jetta Variant                Motor 4.0 Turbo  2003        44410.0    False   
Passat                          Motor Diesel  1991         5712.0    False   
Crossfox                     Motor Diesel V8  1990        37123.0    False   
DS5                          Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4             Motor 2.4 Turbo  2006        25757.0    False   
...                                      ...   ...            ...      ...   
Phantom 2013                        Motor V8  2014        27505.0    False   
Cadillac Ciel concept               Motor V8  1991        29981.0    False   
Classe GLK             Motor 5.0 V8 Bi-Turbo  2002        52637.0    False   
Aston Martin DB5                Motor Diesel  1996         7685.0    False   
Macan                        Motor Diesel V6  1992        50188.0    False   

                                                              Acessórios  \
Nome                                                                       
Jetta Variant          ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox               ['Piloto automático', 'Controle de estabilidad...   
DS5                    ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4       ['Rodas de liga', '4 X 4', 'Central multimídia...   
...                                                                  ...   
Phantom 2013           ['Controle de estabilidade', 'Piloto automátic...   
Cadillac Ciel concept  ['Bancos de couro', 'Painel digital', 'Sensor ...   
Classe GLK             ['Rodas de liga', 'Controle de tração', 'Câmbi...   
Aston Martin DB5       ['Ar condicionado', '4 X 4', 'Câmbio automátic...   
Macan                  ['Central multimídia', 'Teto panorâmico', 'Vid...   

                           Valor  
Nome                              
Jetta Variant           88078.64  
Passat                 106161.94  
Crossfox                72832.16  
DS5                    124549.07  
Aston Martin DB4        92612.10  
...                          ...  
Phantom 2013            51759.58  
Cadillac Ciel concept   51667.06  
Classe GLK              68934.03  
Aston Martin DB5       122110.90  
Macan                   90381.47  

[258 rows x 6 columns]

In [ ]:
dataset.dropna(subset =  ['Quilometragem'], inplace = True)  # deleta todos os registros com valor NaN na variável Quilometragem

In [ ]:
dataset

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                         
Jetta Variant                Motor 4.0 Turbo  2003        44410.0    False   
Passat                          Motor Diesel  1991         5712.0    False   
Crossfox                     Motor Diesel V8  1990        37123.0    False   
Aston Martin DB4             Motor 2.4 Turbo  2006        25757.0    False   
Palio Weekend                  Motor 1.8 16v  2012        10728.0    False   
...                                      ...   ...            ...      ...   
Phantom 2013                        Motor V8  2014        27505.0    False   
Cadillac Ciel concept               Motor V8  1991        29981.0    False   
Classe GLK             Motor 5.0 V8 Bi-Turbo  2002        52637.0    False   
Aston Martin DB5                Motor Diesel  1996         7685.0    False   
Macan                        Motor Diesel V6  1992        50188.0    False   

                                                              Acessórios  \
Nome                                                                       
Jetta Variant          ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox               ['Piloto automático', 'Controle de estabilidad...   
Aston Martin DB4       ['Rodas de liga', '4 X 4', 'Central multimídia...   
Palio Weekend          ['Sensor de estacionamento', 'Teto panorâmico'...   
...                                                                  ...   
Phantom 2013           ['Controle de estabilidade', 'Piloto automátic...   
Cadillac Ciel concept  ['Bancos de couro', 'Painel digital', 'Sensor ...   
Classe GLK             ['Rodas de liga', 'Controle de tração', 'Câmbi...   
Aston Martin DB5       ['Ar condicionado', '4 X 4', 'Câmbio automátic...   
Macan                  ['Central multimídia', 'Teto panorâmico', 'Vid...   

                           Valor  
Nome                              
Jetta Variant           88078.64  
Passat                 106161.94  
Crossfox                72832.16  
Aston Martin DB4        92612.10  
Palio Weekend           97497.73  
...                          ...  
Phantom 2013            51759.58  
Cadillac Ciel concept   51667.06  
Classe GLK              68934.03  
Aston Martin DB5       122110.90  
Macan                   90381.47  

[197 rows x 6 columns]